### PR TITLE
Issue #13 fix - Added multi-threading

### PR DESCRIPTION
**Pull request by amygara**
os_feels

### CODE REVIEW

File: sockets/client-tcp.c [hunk 1]
```diff
@@ -48,6 +48,7 @@ int main(int argc, char ** argv)
                         break;
                 }
         }
+        message = "gr8 b8 m8, r8 8/8";
```
Comment (twood02) [inline sockets/client-tcp.c]: don't modify this file unless there's a good reason

File: threads/client-tcp.c [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+#include <inttypes.h>
+#include <string.h>
+
+/****************************************
+        Author: Tim Wood
+        with a little help from
+        http://beej.us/guide/bgnet/
+****************************************/
+
+int main(int argc, char ** argv)
```
Comment (twood02) [inline threads/client-tcp.c:17]: Is this file identical to the one in sockets?  If so we could just delete it.  I guess it is fine to keep it for now and maybe later we will make this a multi-threaded client.

File: threads/server-tcp.c [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+#include <unistd.h>
+#include <inttypes.h>
+#include <string.h>
+#include <pthread.h>
+
+/****************************************
+        Author: Tim Wood
+        Co-Sub-Authors: Eric Armbrust, Neel Shah, Phil Lopreiato
+        with a little help from
+        http://beej.us/guide/bgnet/
```
Comment (twood02) [inline threads/server-tcp.c:16]: If you have any useful references for threads you can post them here.  "man pthread" is also fine.

File: threads/server-tcp.c [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+                i++;
+                i = i % 1024;
+
+                addr_size = sizeof client_addr;
+                clientfd = accept(sockfd, (struct sockaddr *)&client_addr, &addr_size);
+
+                /* Eric: Create a thread th handle read after server accepts client socket. */
+                pthread_create(&client[i], NULL, (void *)handle_client, (void *)(intptr_t)clientfd);
+
+/******************************
```
Comment (twood02) [inline threads/server-tcp.c]: Just delete all the commented stuff

File: threads/server-tcp.c [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+        /* Time to listen for clients.*/
+        listen(sockfd, BACKLOG);
+        /* Loop forever accepting new connections. */
+        while(1) {
+                int clientfd;
+                socklen_t addr_size;
+                struct sockaddr_storage client_addr;
+
+                i++;
+                i = i % 1024;
```
Comment (twood02) [inline threads/server-tcp.c]: What happens when 513 clients connect to your server sequentially, one after the other?

File: threads/server-tcp.c [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+ * Don't have more than this or bad things *will* happen :c
+ */
+#define MAX_CLIENTS 512
+
+/****************************************
+ * Func passed to pthread_create that handles
+ * the print off post-connection. If more threads
+ * than MAX_CLIENTS call this function, it esplodes
+ * ...don't do it. Serioudsly
+ ****************************************/
```
Comment (twood02) [inline threads/server-tcp.c:39]: let's be more precise: I predict that serious explosions can happen with a concurrency level of 2.  What if one really really slow client connects, then while it is still sending you data 512 other fast but sequential clients connect?

You don't have to fix this, but let's have messages that make it clear what cases will cause the doom.

**Comment by twood02** [comment]
A picture is worth 1000 words, but you still should use a few -- tell me what issue you're fixing and give me a sentence to explain how.  Overall the code looks good--I gave some comments about edge cases.  You can either fix them or give some warning messages in there for others.

File: sockets/client-tcp.c (hunk 1 above)
Comment (amygara) [inline sockets/client-tcp.c]: TBH, I was lazy and didn't want to revert :c

File: threads/server-tcp.c (hunk 5 above)
Comment (amygara) [inline threads/server-tcp.c]: @nks5295 thought it was suuuuuuuuch a good idea to change MAX_CON to 512!!! Need to fix mod.

File: sockets/client-tcp.c (hunk 1 above)
Comment (twood02) [inline sockets/client-tcp.c]: part of our goal here is to discover the horror of shared repos with merging/conflicts, etc. Have fun!

**Comment by amygara** [comment]
Because I can't git: plz merge so I can not fail. Fixes Issue #13

**Pull request merged**